pkg/action/executor/horizontal_scaler.go

```diff
@@ -30,32 +30,63 @@ func (h *HorizontalScaler) Execute(input *TurboActionExecutorInput) (*TurboActio
 		return nil, err
 	}
 	//2. Prepare controllerUpdater
-	controllerUpdater, err := newK8sControllerUpdaterViaPod(h.clusterScraper,
-		pod, h.ormClient, h.gitConfig, h.k8sClusterId)
+	namespace, controllerName, kind, err := getWorkloadControllerInfo(actionItem.GetTargetSE())
+	if err != nil {
+		glog.Errorf("Failed to get controller information: %v", err)
+		return &TurboActionExecutorOutput{}, err
+	}
+	controllerUpdater, err := newK8sControllerUpdater(h.clusterScraper, h.ormClient, kind, controllerName,
+		"", namespace, h.k8sClusterId, nil, h.gitConfig)
 	if err != nil {
 		glog.Errorf("Failed to create controllerUpdater: %v", err)
 		return &TurboActionExecutorOutput{}, err
 	}
 	//3. Execute the action to update replica diff of the controller
+	controllerFullName := util.BuildIdentifier(namespace, controllerName)
 	err = controllerUpdater.updateWithRetry(&controllerSpec{replicasDiff: diff})
 	if err != nil {
-		glog.Errorf("Failed to scale %s: %v", pod.Name, err)
+		glog.Errorf("Failed to scale %s: %v", controllerFullName, err)
 		return &TurboActionExecutorOutput{}, err
 	}
-	podFullName := util.BuildIdentifier(pod.Namespace, pod.Name)
-	glog.V(2).Infof("Action HorizontalScale for pod[%v] succeeded.", podFullName)
+	glog.V(2).Infof("Action HorizontalScale for workload controller[%v] succeeded.", controllerFullName)
 	return &TurboActionExecutorOutput{Succeeded: true}, nil
 }
 
 func getReplicaDiff(action *proto.ActionItemDTO) (int32, error) {
 	atype := action.GetActionType()
-	if atype == proto.ActionItemDTO_PROVISION {
-		// Scale out, increase the replica. diff = 1.
-		return 1, nil
-	} else if atype == proto.ActionItemDTO_SUSPEND {
-		// Scale in, decrease the replica. diff = -1.
-		return -1, nil
-	} else {
-		return 0, fmt.Errorf("action %v is not a scaling action", atype.String())
+	if atype != proto.ActionItemDTO_HORIZONTAL_SCALE {
+		return 0, fmt.Errorf("action %v is not a horizontal scaling action", atype.String())
+	}
+
+	data := action.GetContextData()
+	var oldValue, newValue string
+
+	for _, item := range data {
+		switch *item.ContextKey {
```

Comment:
Please add nil checks here and where ever else you might dereference a pointer value